### ISSUE TITLE
[RFC] Add resource limit for worker without container

### ISSUE THREAD

**Issue by chenk008** [ray-project/ray#17596]
### Describe your feature request

Currently we can limit ray worker process's resource usage when the worker is in a container.  In some scenarios, the container image is not needed, but resource usage limit is needed.

For now the system flag is `worker_resource_limits_enabled`.

## Motivation

- Add resource limit for worker process without container. The resource limit should match scheduler policy.

## Proposal

It is a common practice to put process into specified cgroup to limit resource usage. We can use it to manage ray worker process. 

1. Before starting a worker process, we need to create a cgroup with resource limit config.
2. When the worker process has been forked, it need to add itself into this cgroup.
3. After the worker process  exited, we need to delete the cgroup.
4. Additional:  we can reconfigure the cgroup when the worker process is assigned a new task.

The runtime env agent can be responsible for managing cgroup and rewriting the worker process command.

Related issue:
- https://github.com/ray-project/ray/issues/5114
- https://github.com/ray-project/ray/issues/14077

**Comment by chenk008** [comment]
@eric @edoakes @rkooo567 Please take a look. Would you please let me know your guys' thoughts?

**Comment by ericl** [comment]
I'm in favor of this since it allows memory limits to work whether or not container environments are used (which is a heavier dependency to require). Is it possible to implement this without root access?

**Comment by chenk008** [comment]
@ericl I think we can use `rlimit` on linux to set the process resource limit. It does not need `SYS_ADMIN`.  An unprivileged process may set only its soft limit to a value in the range from 0 up to the hard limit, and (irreversibly) lower its hard limit.

**Comment by chenk008** [comment]
@ericl I think we have some options:

1. Using `rlimit` to set process resource limit on linux. It does not need overloaded capability like `SYS_ADMIN`. Docker has document about it:https://docs.docker.com/engine/security/rootless/#limiting-resources-without-cgroup. But the `RLIMIT_AS` is the maximum size of the process's virtual memory, not the used virtual memory, it is not perfect for memory limit, and `RLIMIT_RSS` does not work.  It's too strict that `RLIMIT_CPU` will send SIGKILL when process' CPU time reached the hard limit.
2. Using `cgroupv2` which can be configure as rootless user and available as of the 4.5 version of the kernel, is now available and supported by most container deployment systems. Rootless containers have become a very popular means to prevent runtime vulnerabilities in containers. 
3. Using `cgroupv1` which is need `SYS_ADMIN` capability. It is overloaded for some cases.

I think the `cgroupv2` is a appropriate approach. WDYT?

**Comment by chenk008** [comment]
@ericl I think we have some options:

1. Using `rlimit` to set process resource limit on linux. It does not need overloaded capability like `SYS_ADMIN`. Docker has document about it:https://docs.docker.com/engine/security/rootless/#limiting-resources-without-cgroup. But the `RLIMIT_AS` is the maximum size of the process's virtual memory, not the used virtual memory, it is not perfect for memory limit, and `RLIMIT_RSS` does not work. And it's too strict that `RLIMIT_CPU` will send SIGKILL when process' CPU time reached the hard limit.

2. `cgroupv2` can be configure as rootless user, it is available as of the 4.5 version of the kernel, and supported by most container deployment systems. Rootless containers have become a very popular means to prevent runtime vulnerabilities in containers.  The ray process can create child cgroup in his own cgroup like `/sys/fs/cgroup/user.slice/user-$(id -u).slice/user@$(id -u).service/user.slice`

3. Using `cgroupv1`. Create a parent cgroup as root, and make it configurable by your user(i.e. ray). https://serverfault.com/questions/478946/how-can-i-create-and-use-linux-cgroups-as-a-non-root-user

I prefer `cgroupv2` because it is more and more widely used. WDYT?

**Comment by ericl** [comment]
Cgroupv2 sounds pretty good to me. Could we add this as an experimental runtime_env plugin? cc @edoakes

**Comment by edoakes** [comment]
I'm fine with making this an experimental runtime_env plugin. @simon-mo could you please shepherd this given that you're working on cleaning up the existing container-related code?

**Comment by chenk008** [comment]
@simon-mo I will clean container-related from `setuo_worker.py` code, make them as a runtime_env plugin. WDYT?

**Comment by simon-mo** [comment]
Sounds good. Please tag me in any PR!

**Comment by simon-mo** [comment]
@chenk008 for the plugin, can you write up the API and user facing documentation? How would user configure the plugin, or no plugin is required at all?

**Comment by simon-mo** [comment]
However this config is global to entire ray cluster. Seems like it's not possible to expose it for per task runtime env without extra config options and handling.

**Comment by chenk008** [comment]
```
For tasks that need isolation and resource guarantee, we can also both set cgroup requests and limits to ensure they get their share right?
```
I think we can both set requests and limits.

@simon-mo It makes sense. So should we add a flag `enable_resource_limit` into runtime_env?

**Comment by SongGuyang** [comment]
Can we restart this discussion and support this feature recently?

**Comment by ericl** [comment]
Seems this was partially merged (at least the flag). What remains to enable this as a feature?

**Comment by SongGuyang** [comment]
@ericl We don't support using cgroup to limit worker processes now, right?

**Comment by SongGuyang** [comment]
The flag `worker_resource_limits_enabled` only works in the case of `worker in container`.

**Comment by ericl** [comment]
Got it. Personally I'm +1 on supporting it. How about we write a REP for it? https://github.com/ray-project/enhancements

IMO, there are a couple different possibilities to consider: 
1. We only support it in worker-in-container (and make it easier to enable worker-in-container: AFAIK, it's not well documented and I haven't heard anyone using it in the wild). It would be awesome if you could just flip a flag and have workers containerized.
2. We support it for both.

**Comment by SongGuyang** [comment]
@ericl Sounds good for using REP. @chenk008 Let's create.

**Comment by jjyao** [comment]
Reopen this since we haven't support per task/actor cgroup yet.
